pkg/counter/cache.go

```diff
@@ -175,24 +220,85 @@ func indexOfMember(members []toolchainv1alpha1.Member, name string) int {
 	return -1
 }
 
-func shouldLoadCurrentResources(toolchainStatus *toolchainv1alpha1.ToolchainStatus) bool {
-	return (toolchainStatus.Status.HostOperator == nil ||
-		toolchainStatus.Status.HostOperator.MasterUserRecordCount == 0) &&
-		!cachedCounts.initialized
+func initialize(cl client.Client, toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
+	// skip if cached counters are already initialized
+	if cachedCounts.initialized {
+		return nil
+	}
+
+	// initialize the cached counters from the UserSignup and MasterUserRecord resources.
+	if toolchainStatus.Status.HostOperator == nil ||
+		toolchainStatus.Status.HostOperator.MasterUserRecordCount == 0 {
+		return initializeFromResources(cl, toolchainStatus.Namespace)
+	}
+	// otherwise, initialize the cached counters from the ToolchainStatud resource.
+	return initializeFromToolchainStatus(cl, toolchainStatus)
 }
 
-func loadCurrentResources(cl client.Client, namespace string) error {
+// initialize the cached counters from the UserSignup and MasterUserRecord resources.
+// this func lists all UserSignup and MasterUserRecord resources
+func initializeFromResources(cl client.Client, namespace string) error {
+	log.Info("initializing counters from resources")
+	usersignups := &toolchainv1alpha1.UserSignupList{}
+	if err := cl.List(context.TODO(), usersignups, client.InNamespace(namespace)); err != nil {
+		return err
+	}
 	murs := &toolchainv1alpha1.MasterUserRecordList{}
 	if err := cl.List(context.TODO(), murs, client.InNamespace(namespace)); err != nil {
 		return err
 	}
 	reset()
+	for _, usersignup := range usersignups.Items {
+		if value, exists := usersignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey]; exists {
+			_, err := strconv.Atoi(value) // let's make sure the value is actually an integer
+			if err != nil {
+				log.Error(err, "invalid number of activations", "name", usersignup.Name, "value", value)
+				continue
+			}
+			cachedCounts.UsersPerActivationCounts[value]++
+		} else {
+			// annotation wasn't already set by the UserSignup controller at the time where the list was loaded,
+			// so let's assume the value is '1'
+			cachedCounts.UsersPerActivationCounts["1"]++
```

Comment:
We should ignore such usersignups. There are signups which are not active (pending verification for example). We should not count them.

Reply:
I agree that I should excluded pending users, but wouldn't it be safe to count them here as well? for example, if the counter initialization happens before the UserSignup controller had time to set the annotation on all resources?

Reply:
If the counter is initialized before the UserSignup controller sets the annotations, then it should still count only those who have the annotation - the UserSignup controller will finish the job later and counts the missing UserSignups when it adds the missing annotation as was proposed here https://github.com/codeready-toolchain/host-operator/pull/401#discussion_r609484476

Reply:
good point, so I'll ignore these "unannotated" UserSignups then

Reply:
done in c795693